### PR TITLE
Expanded and Improved Catalog Customization Docs

### LINKED ISSUE #19376: 🚀 Feature: Catalog Customization Docs Improvements
### 🔖 Feature description

The current Catalog Customization documentation is more like a tutorial for adding a custom filter. This is very helpful but it's a bit of a deep dive to begin with. We should revise this to start off with some of the more basic customization building up to creating a custom filter. 

### 🎤 Context

This often comes up on Discord and the answer is often to past code examples. If this was documented then people might be able to find it on their own or we can send them to this page providing a consistent response

### ✌️ Possible Implementation

Revise the page to include how to use the following:

- initially selected filter
- adding/modifying columns
- adding/modifying actions
- initial kind
- how to handle empty content
- setting owner picker mode
- using the table options

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

### PR DESCRIPTION
**Pull request by awanlin**
## Hey, I just made a Pull Request!

Expanded and improved Catalog Customization documentation. Also, added support for passing in custom filters to `CatalogIndexPage`.

Closes #19376

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx [hunk 1]
```diff
@@ -103,6 +103,7 @@ export interface DefaultCatalogPageProps {
   emptyContent?: ReactNode;
   ownerPickerMode?: EntityOwnerPickerProps['mode'];
   pagination?: boolean | { limit?: number };
+  filters?: ReactNode;
```
Comment (awanlin) [inline plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx:106]: Though this works it does mean having to pass in all the filters. If there are suggestions to make this work where you can just add on but also replace the existing filters I'm very open to making those changes

**Comment by awanlin** [comment]
If this related PR https://github.com/backstage/backstage/pull/22863 gets merged in before this one then we'll need to refactor this PR to include it.

File: docs/features/software-catalog/catalog-customization.md [hunk 2]
```diff
@@ -5,64 +5,159 @@ title: Catalog Customization
...
+  path="/catalog"
+  element={<CatalogIndexPage initiallySelectedFilter="all" />}
+/>
+```
+
+Possible options are: owned, starred, or all
+
+## Initially Selected Kind
+
+By default the initially selected Kind when viewing the Catalog in Component, but you may have reasons that you want this to be different. Let's say at your Organization they would like it to always default to Domain, here's how you would do that:
```
Comment (adamdmharvey) [inline docs/features/software-catalog/catalog-customization.md:33]: Typo: "is component" in first sentence

File: plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx (hunk 1 above)
Comment (freben) [reply plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx:106]: Were you looking at making a DefaultFilters to go with this, to support extension of the defaults?

File: docs/features/software-catalog/catalog-customization.md [hunk 3]
```diff
@@ -5,64 +5,159 @@ title: Catalog Customization
...
+
+```tsx title="packages/app/src/App.tsx"
+<Route path="/catalog" element={<CatalogIndexPage ownerPickerMode="all" />} />
+```
+
+Possible options are: owners-only or all
+
+## Customize Columns
+
+By default the columns you see in the `CatalogIndexPage` were selected to be a good starting point for most but there may be reasons that you would like to customize these with more or less columns. On primary use case for this customization is if you added a custom Kind. Support for this was added in v1.23.0 of Backstage, make sure you are on that version or newer to use this feature. Here's an example of how to make this customization:
```
Comment (adamdmharvey) [inline docs/features/software-catalog/catalog-customization.md:53]:
```suggestion
By default the columns you see in the `CatalogIndexPage` were selected to be a good starting point for most but there may be reasons that you would like to customize these with more or less columns. One primary use case for this customization is if you added a custom Kind. Support for this was added in v1.23.0 of Backstage, make sure you are on that version or newer to use this feature. Here's an example of how to make this customization:
```

File: docs/features/software-catalog/catalog-customization.md [hunk 4]
```diff
@@ -5,64 +5,159 @@ title: Catalog Customization
...
-`catalog-backend`, and a way to hook in filters.
+> Note: the above example has been simplified and you will most likely have more code then just this in your `App.tsx` file.
 
-Now we're ready to create a new filter that implements the `EntityFilter`
-interface:
+## Customize Actions
+
+The `CatalogIndexPage` comes with three default actions - view, edit, and star. You might want to add more here's how:
+
+First you'll need to add `@mui/utils` to your `packages/app/package.json`:
```
Comment (adamdmharvey) [inline docs/features/software-catalog/catalog-customization.md:91]:
```suggestion
The `CatalogIndexPage` comes with three default actions - view, edit, and star. You might want to add more.

To do this, first you'll need to add `@mui/utils` to your `packages/app/package.json`:
```

Possible suggestion, just removes the duplicate colon entry points.

File: docs/features/software-catalog/catalog-customization.md [hunk 5]
```diff
@@ -5,64 +5,159 @@ title: Catalog Customization
...
+<Route path="/catalog" element={<CatalogIndexPage actions={customActions} />} />
+```
+
+> Note: the above example has been simplified and you will most likely have more code then just this in your `App.tsx` file.
+
+The above customization will override the existing actions. Currently the only way to keep them and add your own is to also include the existing actions in your array by copying them from the [`defaultActions`](https://github.com/backstage/backstage/blob/57397e7d6d2d725712c439f4ab93f2ac6aa27bf8/plugins/catalog/src/components/CatalogTable/CatalogTable.tsx#L113-L168).
+
+## Custom Filters
+
+You can add custom filters. For example, suppose that I want to allow filtering by a custom annotation added to entities, `company.com/security-tier`. Here is how we can built a filter to support that need.
```
Comment (adamdmharvey) [inline docs/features/software-catalog/catalog-customization.md:158]:
```suggestion
You can add custom filters. For example, suppose that I want to allow filtering by a custom annotation added to entities, `company.com/security-tier`. Here is how we can build a filter to support that need.
```

**Review by adamdmharvey — COMMENTED** [review]
Few more; this documentation is looking great @awanlin !

File: plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx (hunk 1 above)
Comment (awanlin) [reply plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx:106]: Yes, plan to tackle that tomorrow.

File: docs/features/software-catalog/catalog-customization.md (hunk 2 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md]: Good catch, fixed 👍

File: docs/features/software-catalog/catalog-customization.md (hunk 3 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md]: Good catch, fixed 👍

File: docs/features/software-catalog/catalog-customization.md (hunk 4 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md]: I like it, changed to match 👍

File: plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx (hunk 1 above)
Comment (awanlin) [reply plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx:99]: Brain wasn't working yesterday but I got this done now, hopefully this was what you had in mind?

**Comment by awanlin** [comment]
@adamdmharvey, any chance you can take another look? Had to reformat some of the docs as I introduced a new `DefaultFilters` component which caused a few changes. No rush, when you have time. 👍

**Comment by freben** [comment]
Seems promising, but it also seems to have triggered some form of CI error!

File: plugins/catalog-react/src/components/DefaultFilters/DefaultFilters.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  EntityTypePicker,
+  UserListPicker,
+  EntityOwnerPicker,
+  EntityLifecyclePicker,
+  EntityTagPicker,
+  EntityProcessingStatusPicker,
+  EntityNamespacePicker,
+  UserListFilterKind,
+  EntityOwnerPickerProps,
+} from '@backstage/plugin-catalog-react';
```
Comment (freben) [inline plugins/catalog-react/src/components/DefaultFilters/DefaultFilters.tsx:28]: these should be local paths right? it's all from inside the same package it seems

Comment (awanlin) [reply plugins/catalog-react/src/components/DefaultFilters/DefaultFilters.tsx:28]: Yes, when I created the `DefaultFilters` I did that in `@backstage/plugin-catalog` and thought it made more sense to be in `@backstage/plugin-catalog-react`. I'll fix these up 👍

Comment (awanlin) [reply plugins/catalog-react/src/components/DefaultFilters/DefaultFilters.tsx]: Imports have been fixed 👍

**Comment by awanlin** [comment]
@freben looks like fixing the imports may have also fixed the CI, it's green now 👍

File: docs/features/software-catalog/catalog-customization.md [hunk 7]
```diff
@@ -5,64 +5,232 @@ title: Catalog Customization
...
-If you want to change the default index page - such as to add a custom filter to
-the catalog - you can create your own `CatalogIndexPage`.
+## Pagination
 
-> Note: The catalog index page is designed to have a minimal code footprint to
-> support easy customization, but creating a copy does introduce a possibility
-> of drifting out of date over time. Be sure to check the catalog
-> [CHANGELOG](https://github.com/backstage/backstage/blob/master/plugins/catalog/CHANGELOG.md)
-> periodically.
+Initial support for pagination of the `CatalogIndexPage` was added in v1.21.0 of Backstage, make sure you are on that version or newer to use this feature. To enable pagination you simply need to pass in the `pagination` prop like this:
```
Comment (adamdmharvey) [inline docs/features/software-catalog/catalog-customization.md:12]:
```suggestion
Initial support for pagination of the `CatalogIndexPage` was added in v1.21.0 of Backstage, so make sure you are on that version or newer to use this feature. To enable pagination you simply need to pass in the `pagination` prop like this:
```

File: docs/features/software-catalog/catalog-customization.md [hunk 8]
```diff
@@ -5,64 +5,232 @@ title: Catalog Customization
...
+<Route path="/catalog" element={<CatalogIndexPage actions={customActions} />} />
+```
+
+> Note: the above example has been simplified and you will most likely have more code then just this in your `App.tsx` file.
+
+The above customization will override the existing actions. Currently the only way to keep them and add your own is to also include the existing actions in your array by copying them from the [`defaultActions`](https://github.com/backstage/backstage/blob/57397e7d6d2d725712c439f4ab93f2ac6aa27bf8/plugins/catalog/src/components/CatalogTable/CatalogTable.tsx#L113-L168).
+
+## Customize Filters
+
+There are tree options you have for filters: adjusting the existing filters with props, adding or removing the default filters, or creating a brand new custom filter. The following sections cover these cases
```
Comment (adamdmharvey) [inline docs/features/software-catalog/catalog-customization.md:171]:
```suggestion
There are three options you have for filters: adjusting the existing filters with props, adding or removing the default filters, or creating a brand new custom filter. The following sections cover these cases
```

🌴 ? 😁

File: docs/features/software-catalog/catalog-customization.md [hunk 9]
```diff
@@ -5,64 +5,232 @@ title: Catalog Customization
...
+
+### Default Filter Props
+
+There is a set of default filters that you can use, they surface all the props mentioned earlier in this document. Here's how they can be used:
```
Comment (adamdmharvey) [inline docs/features/software-catalog/catalog-customization.md:175]:
```suggestion
There are a set of default filters that you can use, which surface all the props mentioned earlier in this document. Here's how they can be used:
```

File: docs/features/software-catalog/catalog-customization.md [hunk 10]
```diff
@@ -48,6 +48,19 @@ The Owner filter by default will only contain a list of Users and/or Groups that
 
 Possible options are: owners-only or all
 
+## Table Options
+
+The tables used within Backstage are built on top of [`@material-table/core`](https://material-table-core.github.io/) and the `CatalogIndexPage` has a `tableOptions` prop that allows you to customize the underlying table to a certain extent, there are some hard coded Backstage settings that can't be changed. Here's an example of how to use this prop to disable the search filter field in the table's header:
```
[the remark this answers is not in the document]
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md]: Fixed 👍

File: docs/features/software-catalog/catalog-customization.md (hunk 7 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md]: Fixed 👍

File: docs/features/software-catalog/catalog-customization.md (hunk 8 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md]: LOL! Fixed 👍

File: docs/features/software-catalog/catalog-customization.md (hunk 9 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md]: Fixed 👍

File: plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx [hunk 11]
```diff
@@ -115,21 +109,19 @@ export function DefaultCatalogPage(props: DefaultCatalogPageProps) {
...
-          <EntityTypePicker />
-          <UserListPicker initialFilter={initiallySelectedFilter} />
-          <EntityOwnerPicker mode={ownerPickerMode} />
-          <EntityLifecyclePicker />
-          <EntityTagPicker />
-          <EntityProcessingStatusPicker />
-          <EntityNamespacePicker />
-        </>
+        filters ?? (
+          <DefaultFilters
```
Comment (freben) [inline plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx:119]: There's a mild risk of confusion that if you pass in custom filters, the `initialKind`, `initiallySelectedFilter` and `ownerPickerMode` props will have no effect. But it's probably ok

File: docs/features/software-catalog/catalog-customization.md [hunk 12]
```diff
@@ -140,53 +300,40 @@ export const EntitySecurityTierPicker = () => {
...
+Now we can add the component to `CatalogIndexPage`:
 
 ```tsx title="packages/app/src/App.tsx"
+{
+  /* highlight-add-start */
+}
+import { DefaultFilters } from '@backstage/plugin-catalog-react';
+{
+  /* highlight-add-end */
+}
```
Comment (freben) [inline docs/features/software-catalog/catalog-customization.md:312]:
```suggestion
{/* highlight-add-start */}
import { DefaultFilters } from '@backstage/plugin-catalog-react';
{/* highlight-add-end */}
```

File: docs/features/software-catalog/catalog-customization.md [hunk 13]
```diff
@@ -5,64 +5,232 @@ title: Catalog Customization
...
-to entities, `company.com/security-tier`. To start, I'll copy the code for the
-default catalog page and create a component.
+```tsx title="packages/app/src/App.tsx"
+<Route path="/catalog" element={<CatalogIndexPage pagination />} />
+```
 
-```tsx
-// imports, etc omitted for brevity. for full source see:
-// https://github.com/backstage/backstage/blob/master/plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx
-export const CustomCatalogPage = ({
```
Comment (freben) [inline docs/features/software-catalog/catalog-customization.md:28]: On this deletion, I'll say though, I think the original idea was that it opens up for really ANY type of customization fairly easily (at the cost of these lines of boilerplate). Like replacing the header and what's in it,  the support button ...

Might be worth retaining this as the advanced option, such that "if you need to go beyond these simpler options, here's how to replace the page as a whole using its constituent sub-parts" so to speak

**Review by freben — APPROVED** [review]
Alright, looking good.

Pulling in @backstage/maintainers and @backstage/catalog-maintainers for blessing of the concept of using `filters?: ReactNode` as the input

**Review by mikeyhc — APPROVED** [review]

**Comment by GoFightNguyen** [comment]
I really like the updated `catalog-customization.md` :chefskiss:

File: docs/features/software-catalog/catalog-customization.md (hunk 13 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md:28]: Let me see what I can do to add this back, it's a valid option for those who want to go much deeper customizing htis. The problem I was running into was that it wasn't as simple as copying what we as at - https://github.com/backstage/backstage/blob/master/plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx - I needed to pull in more "things".

File: docs/features/software-catalog/catalog-customization.md (hunk 12 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md:312]: Fixed, seems every time my editor applies the formatting it does this, not sure why just here and not the other cases 🤔

File: docs/features/software-catalog/catalog-customization.md (hunk 13 above)
Comment (awanlin) [reply docs/features/software-catalog/catalog-customization.md:28]: Alright, added back but had to do s bit of adjustments, think this worked enough to give people the right idea.

**Review by freben — APPROVED** [review]

**Comment by awanlin** [comment]
Thanks @adamdmharvey @freben @mikeyhc 🚀

**Pull request merged**